crates/cli/src/subcommands/sql.rs

```diff
@@ -76,50 +91,237 @@ pub(crate) async fn parse_req(mut config: Config, args: &ArgMatches) -> Result<C
+    ///   }]
+    /// }
+    /// {
+    ///   "rows": [{
+    ///     "identity": [
+    ///       "aba52919637a60eb336e76eed40843653bfb3d9d94881d78ab13dda56363b993"
+    ///     ],
+    ///     "name": {
+    ///       "1": []
+    ///     },
+    ///     "online": true
+    ///   }]
+    /// }
+    /// ```
+    ///
+    /// [json-seq]: https://datatracker.ietf.org/doc/html/rfc7464
+    async fn render_json(&self, mut out: impl AsyncWrite + Unpin) -> anyhow::Result<()> {
+        for stmt_result in &self.stmt_results {
+            let client_api::StmtResultJson { schema, rows } = stmt_result.try_into()?;
 
-    let stmt_result_json: Vec<StmtResultJson> = serde_json::from_str(&json)?;
+            let col_names = Vec::from(schema.elements)
+                .into_iter()
+                .enumerate()
+                .map(|(pos, col)| col.name.unwrap_or(pos.to_string().into()))
+                .collect::<Vec<_>>();
+            let rows = rows
+                .into_iter()
+                .map(|row| col_names.iter().zip(Vec::from(row.elements)).collect::<HashMap<_, _>>())
+                .collect::<Vec<_>>();
 
-    // Print only `OK for empty tables as it's likely a command like `INSERT`.
-    if stmt_result_json.is_empty() {
-        if with_stats {
-            print_timings(now);
+            format::write_json_seq(&mut out, &json!({"rows": rows})).await?;
         }
-        println!("OK");
-        return Ok(());
-    };
+        out.flush().await?;
 
-    stmt_result_json
-        .iter()
-        .map(|stmt_result| {
+        Ok(())
+    }
+
+    /// Renders each query result as an ASCII table similar in style to `psql`.
+    ///
+    /// If `with_timing` is given, timing information about the whole request
+    /// is printed after all tables have been rendered.
+    ///
+    /// If `with_row_count` is given, the number of rows is printed in the
+    /// footer of each table.
+    ///
+    /// # Example
+    ///
+    /// A query `SELECT * FROM Message; SELECT * FROM Users;` against the
+    /// `quickstart-chat` application may return the following:
+    ///
+    /// ```text
+    ///  sender                                                             | sent             | text
+    /// --------------------------------------------------------------------+------------------+----------------------
+    ///  0xaba52919637a60eb336e76eed40843653bfb3d9d94881d78ab13dda56363b993 | 1718611618273455 | "hello"
+    ///
+    ///  identity                                                           | name        | online
+    /// --------------------------------------------------------------------+-------------+--------
+    ///  0xaba52919637a60eb336e76eed40843653bfb3d9d94881d78ab13dda56363b993 | (none = ()) | true
+    /// ```
+    async fn render_tabled(
+        &self,
+        mut out: impl AsyncWrite + Unpin,
+        with_timing: bool,
+        with_row_count: bool,
+    ) -> anyhow::Result<()> {
+        if self.stmt_results.is_empty() {
```

Comment:
Note that this is actually not great on a more fundamental level: nothing prevents you from submitting a mix of select and insert statements, but we won’t print OK for each insert result, because there is none in the response data.